Source/WebKit/Shared/TextExtractionToStringConversion.cpp

```diff
@@ -0,0 +1,411 @@
+        auto index = m_nextLineIndex++;
+        m_lines.resize(m_nextLineIndex);
+        return index;
+    }
+
+private:
+    Vector<String> m_lines;
+    unsigned m_nextLineIndex { 0 };
+    CompletionHandler<void(String&&)> m_completion;
+};
+
+static String commaSeparatedString(const Vector<String>& parts)
+{
+    return makeStringByJoining(parts, ","_s);
+}
+
+static String escapeString(const String& string)
+{
+    auto result = string;
+    result = makeStringByReplacingAll(result, '\\', "\\\\"_s);
+    result = makeStringByReplacingAll(result, '\n', "\\n"_s);
+    result = makeStringByReplacingAll(result, '\r', "\\r"_s);
+    result = makeStringByReplacingAll(result, '\t', "\\t"_s);
+    result = makeStringByReplacingAll(result, '\'', "\\'"_s);
+    result = makeStringByReplacingAll(result, '"', "\\\""_s);
+    result = makeStringByReplacingAll(result, '\0', "\\0"_s);
+    result = makeStringByReplacingAll(result, '\b', "\\b"_s);
+    result = makeStringByReplacingAll(result, '\f', "\\f"_s);
+    result = makeStringByReplacingAll(result, '\v', "\\v"_s);
+    return result;
+}
+
+static Vector<String> eventListenerTypesToStringArray(OptionSet<TextExtraction::EventListenerCategory> eventListeners)
+{
+    Vector<String> result;
+    if (eventListeners.contains(TextExtraction::EventListenerCategory::Click))
+        result.append("click"_s);
+    if (eventListeners.contains(TextExtraction::EventListenerCategory::Hover))
+        result.append("hover"_s);
+    if (eventListeners.contains(TextExtraction::EventListenerCategory::Touch))
+        result.append("touch"_s);
+    if (eventListeners.contains(TextExtraction::EventListenerCategory::Wheel))
+        result.append("wheel"_s);
+    if (eventListeners.contains(TextExtraction::EventListenerCategory::Keyboard))
+        result.append("keyboard"_s);
+    return result;
+}
+
+static Vector<String> partsForItem(const TextExtraction::Item& item)
+{
+    Vector<String> parts;
+
+    if (item.nodeIdentifier)
+        parts.append(makeString("uid="_s, item.nodeIdentifier->toUInt64()));
+
+    if (item.children.isEmpty()) {
+        auto origin = item.rectInRootView.location();
+        auto size = item.rectInRootView.size();
+        parts.append(makeString("["_s,
+            static_cast<int>(origin.x()), ',', static_cast<int>(origin.y()), ";"_s,
+            static_cast<int>(size.width()), 'x', static_cast<int>(size.height()), ']'));
+    }
+
+    if (!item.accessibilityRole.isEmpty())
+        parts.append(makeString("role='"_s, escapeString(item.accessibilityRole), '\''));
+
+    auto listeners = eventListenerTypesToStringArray(item.eventListeners);
+    if (!listeners.isEmpty())
+        parts.append(makeString("events=["_s, commaSeparatedString(listeners), ']'));
+
+    auto attributeKeys = copyToVector(item.ariaAttributes.keys());
+    std::sort(attributeKeys.begin(), attributeKeys.end(), [](const String& a, const String& b) {
+        return codePointCompareLessThan(a, b);
+    });
```

Comment:
Slightly less typing

```suggestion
    std::ranges::sort(attributeKeys, codePointCompareLessThan);
```